allennlp/interpret/attackers/hotflip.py

```diff
@@ -0,0 +1,203 @@
+        """
+        original_instances = self.predictor.inputs_to_labeled_instances(inputs)
+        original_tokens = list(original_instances[0][name_of_input_field_to_attack].tokens)
+        final_tokens = []
+        new_prediction = None
+        for new_instances in original_instances:
+            new_instances = [new_instances]
+            test_instances = self.predictor.inputs_to_labeled_instances(inputs)
+
+            # get a list of fields that we want to check to see if they change
+            # (we want to change model predictions)
+            fields_to_compare = {}
+            for key in new_instances[0].fields:
+                if key not in inputs.keys() and key != name_of_input_field_to_attack:
+                    fields_to_compare[key] = test_instances[0][key]
+
+            current_tokens = new_instances[0][name_of_input_field_to_attack].tokens
+            grads, outputs = self.predictor.get_gradients(new_instances)
+            flipped = []
+            while True:
+                # Compute L2 norm of all grads.
+                grad = grads[name_of_grad_input_field]
+                grads_mag = [numpy.sqrt(g.dot(g)) for g in grad]
+
+                # only flip a token once
+                for index in flipped:
+                    grads_mag[index] = -1
+
+                # we flip the token with highest gradient norm
+                index_of_token_to_flip = numpy.argmax(grads_mag)
+                if grads_mag[index_of_token_to_flip] == -1:
+                    break
+                flipped.append(index_of_token_to_flip)
+
+                # Get new token using taylor approximation
+                input_tokens = new_instances[0][name_of_input_field_to_attack]._indexed_tokens["tokens"]
+                original_id_of_token_to_flip = input_tokens[index_of_token_to_flip]
+                new_id_of_flipped_token = \
+                    first_order_taylor(grad[index_of_token_to_flip],
+                                       self.token_embedding.weight,
+                                       original_id_of_token_to_flip)
+                # flip token
+                new_instances[0][name_of_input_field_to_attack].tokens[index_of_token_to_flip] = \
+                    Token(self.vocab._index_to_token["tokens"][new_id_of_flipped_token])
+                new_instances[0].indexed = False
+
+                # Get model predictions on new_instances, and then label the instances
+                grads, outputs = self.predictor.get_gradients(new_instances) # predictions
+                for key in outputs:
+                    if isinstance(outputs[key], torch.Tensor):
+                        outputs[key] = outputs[key].detach().cpu().numpy().squeeze().squeeze()
+                    elif isinstance(outputs[key], list):
+                        outputs[key] = outputs[key][0]
+                # add labels to new_instances
+                self.predictor.predictions_to_labeled_instances(new_instances[0], outputs)
```

Comment:
Wait, what?  Shouldn't this be taking a return value?  The way this is written, it assumes that you're modifying `new_instance`.  Only some of the implementations do that, and I would say that _none_ of them should.  This needs to be fixed.